### PR TITLE
GODRIVER-2521 Add support for Azure IMDS.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-2521

## Summary
Support the Azure VM-assigned Managed Identity for Automatic KMS Credentials.

## Background & Motivation
Reused the AWS SDK interface to add the support of Azure credential provider.

specs:
https://github.com/mongodb/specifications/blob/d6b8cce6abb3b8e1a0b8f1dc7ee737e18322cfce/source/client-side-encryption/client-side-encryption.rst#obtaining-an-access-token-for-azure-key-vault

tests:
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#azure-imds-credentials
https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/README.rst#azure-imds-credentials-integration-test

### CODE REVIEW

File: .evergreen/config.yml [hunk 1]
```diff
@@ -1013,6 +1013,21 @@ functions:
...
+        script: |
+          ${PREPARE_SHELL}
+          cd ${DRIVERS_TOOLS}/.evergreen/csfle
+          . ./activate-kmstlsvenv.sh
+    - command: shell.exec
+      params:
+        background: true
+        script: |
+          cd ${DRIVERS_TOOLS}/.evergreen/csfle
+          ./kmstlsvenv/bin/python3 bottle.py fake_azure:imds
```
Comment (kevinAlbs) [inline .evergreen/config.yml:1029]: Suggest adding `await_server "Azure" 8080` to `run-tests.sh` here: https://github.com/mongodb/mongo-go-driver/blob/eb83f20dc02a797b03aa1b4b988f98dd4fb0bd9d/.evergreen/run-tests.sh#L85

Using `background: true` will run this shell script in the background. Using `await_server` may prevent a race between the mock server starting and the test running.

File: .evergreen/config.yml [hunk 2]
```diff
@@ -1096,6 +1111,28 @@ functions:
           PKG_CONFIG_PATH=$PKG_CONFIG_PATH \
           LD_LIBRARY_PATH=$LD_LIBRARY_PATH
 
+  run-azure-imds-test:
```
Comment (kevinAlbs) [inline .evergreen/config.yml:1114]: I expect the Azure mock server tests do not need to run in a separate Evergreen task.

Consider combining this into the tests run in run-tests.sh:

- Add the commands from `start-kms-mock-azure-imds-server` to `start-cse-servers`. `start-cse-servers` is run before every task as part of the `pre` tasks.
- Remove the `run-azure-imds-test` task.
- Add `export AZURE_MOCK_SERVERS_RUNNING=true` to the `run-tests.sh` script.
- Remove the `evg-test-azure` Makefile target.

File: .evergreen/config.yml [hunk 3]
```diff
@@ -2411,6 +2527,45 @@ task_groups:
...
+            export AZUREKMS_PUBLICKEYPATH=/tmp/testazurekms.pubkey
+            echo '${testazurekms_privatekey}' > /tmp/testazurekms.prikey
+            sudo chmod 600 /tmp/testazurekms.prikey
+            export AZUREKMS_PRIVATEKEYPATH=/tmp/testazurekms.prikey
+            export AZUREKMS_SCOPE=${AZUREKMS_SCOPE}
+            $DRIVERS_TOOLS/.evergreen/csfle/azurekms/create-and-setup-vm.sh
+      - command: expansions.update
+        params:
+          file: testazurekms-expansions.yml
+    teardown_group:
```
Comment (kevinAlbs) [inline .evergreen/config.yml:2559]:
```suggestion
    teardown_group:
      # setup_group may have failed before updating expansions. Try to update again to clean up resources on setup failure.
      - command: expansions.update
        params:
          ignore_missing_file: true
          file: testazurekms-expansions.yml
```

If the `setup_group` fails, the `expansions.update` may not have been run. That may result in resources not being destroyed.

File: x/mongo/driver/mongocrypt/mongocrypt.go
Comment (kevinAlbs) [inline x/mongo/driver/mongocrypt/mongocrypt.go:1]: Suggest updating the comment for `GetKmsProviders` to clarify it obtains credentials from the environment:

```
// GetKmsProviders attempts to obtain credentials from environment. It is expected to be called when a libmongocrypt context is in the mongocrypt.NeedKmsCredentials state.
```

File: x/mongo/driver/auth/creds/azurecreds.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package creds
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"go.mongodb.org/mongo-driver/internal/aws/credentials"
```
Comment (kevinAlbs) [inline x/mongo/driver/auth/creds/azurecreds.go:14]: Suggest adding a comment in `internal/aws/credentials.go` to note it is now also used to represent Azure credentials:

```
// Also used to fetch Azure credentials from the environment. Azure credentials only consist of an access token. The access token is stored in the `SessionToken` field.
```

**Review by kevinAlbs — COMMENTED** [review]
Looks good. Left suggestions to simplify Evergreen configuration.

File: cmd/testawskms/main.go
Comment (matthewdale) [inline cmd/testawskms/main.go:1]: This package is technically exported. It doesn't export any symbols, so it seems like this should be safe to move, but we've previously avoided moving `cmd/` packages to prevent any risk of imports breaking.

1. Is moving this package a breaking change?
2. If it's not a breaking change, should we move the new package to an `internal` directory to prevent future concerns about moving exported packages?

File: cmd/testkms/main.go [hunk 5]
```diff
@@ -13,20 +13,61 @@ import (
...
+	},
+	"gcp": bson.M{
+		"projectId": "devprod-drivers",
+		"location":  "global",
+		"keyRing":   "key-ring-csfle",
+		"keyName":   "key-name-csfle",
+	},
+}
+
+var providers = func() []string {
```
Comment (matthewdale) [inline cmd/testkms/main.go:38]: Optional: Consider moving this variable declaration into the `if !ok` block in `main` since that's the only place it's used.

E.g.
```go
func main() {
	// ...
	if !ok {
		// ...
		providers := make([]string, 0, len(datakeyopts))
		for p := range datakeyopts {
			providers = append(providers, p)
		}
		fmt.Println("- PROVIDER as a KMS provider, which supports:", strings.Join(providers, ", "))
		// ...
```

File: cmd/testawskms/main.go
Comment (qingyang-hu) [reply cmd/testawskms/main.go:1]: I did not get the motivation for the `cmd` package. All files inside seem to be used for tests. Maybe we can move it to an `internal` directory entirely?

File: .evergreen/config.yml (hunk 2 above)
Comment (kevinAlbs) [reply .evergreen/config.yml]: > Add the commands from start-kms-mock-azure-imds-server to start-cse-servers. start-cse-servers is run before every task as part of the pre tasks.

a428b6a39607290b2dd4389a7c363e79830088a8 adds the `start-kms-mock-azure-imds-server` back. Does adding to `start-cse-servers` not work as expected?

**Review by kevinAlbs — APPROVED** [review]
LGTM. I think `start-kms-mock-azure-imds-server` can be merged into `start-cse-servers` but do not think that should block merging.

File: cmd/testawskms/main.go
Comment (matthewdale) [reply cmd/testawskms/main.go:1]: The motivation for the `cmd` directory is that it's a conventional way to organize Go build targets (i.e. Go `main` packages that are used to build Go programs). The downside is that it's visible in the Go docs (e.g. [here](https://pkg.go.dev/go.mongodb.org/mongo-driver@v1.11.4#section-directories)), but the `internal` directory is hidden.

It turns out if you try to import a `main` package to another project, you get a build error like:
> import "go.mongodb.org/mongo-driver/cmd/testawskms" is a program, not an importable package

As a result, customers actually can't import the `cmd/testawskms` package, even if it exports symbols, so moving the package should be safe. Maybe eventually we want to move it to `internal/` (see [GODRIVER-2724](https://jira.mongodb.org/browse/GODRIVER-2724)), but that isn't a high priority right now.

TLDR: The change is safe as-is and there's no need to move anything to `internal/` right now.

File: internal/credproviders/imds_provider.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		expiryWindow: expiryWindow,
+	}
+}
+
+// RetrieveWithContext retrieves the keys from the Azure service.
+func (a *AzureProvider) RetrieveWithContext(ctx context.Context) (credentials.Value, error) {
+	v := credentials.Value{ProviderName: AzureProviderName}
+	req, err := http.NewRequest(http.MethodGet, azureURI, nil)
+	if err != nil {
+		return v, internal.WrapErrorf(err, "unable to retrieve Azure credentials")
```
Comment (matthewdale) [inline internal/credproviders/imds_provider.go:50]: Optional: Now that the Go driver supports Go 1.13+, `internal.WrapErrorf` can be replaced with `fmt.Errorf` using the `%w` verb.

E.g.
```go
return v, fmt.Errorf("unable to retrieve Azure credentials: %w", err)
```

That suggestion applies to all usages of `internal.WrapErrorf`.

File: mongo/integration/client_side_encryption_prose_test.go [hunk 7]
```diff
@@ -2022,6 +2025,100 @@ func TestClientSideEncryptionProse(t *testing.T) {
 		})
 	})
 
+	mt.RunOpts("18. Azure IMDS Credentials", noClientOpts, func(mt *mtest.T) {
+		buf := make([]byte, 0, 256)
+		kmsProvidersMap := map[string]map[string]interface{}{
+			"azure": {},
+		}
+		p, err := bson.MarshalAppendWithRegistry(bson.DefaultRegistry, buf[:0], kmsProvidersMap)
```
Comment (matthewdale) [inline mongo/integration/client_side_encryption_prose_test.go:2033]: Optional: `MarshalAppendWithRegistry` will be removed in Go Driver 2.0. Since you're using `bson.DefaultRegistry`, consider using `bson.MarshalAppend` instead.

```suggestion
		p, err := bson.MarshalAppend(buf[:0], kmsProvidersMap)
```

File: mongo/integration/client_side_encryption_prose_test.go [hunk 8]
```diff
@@ -2022,6 +2025,100 @@ func TestClientSideEncryptionProse(t *testing.T) {
...
+		assert.Nil(mt, err, "error in MarshalAppendWithRegistry: %v", err)
```
Comment (matthewdale) [inline mongo/integration/client_side_encryption_prose_test.go:2034]: Optional: If an error here would make the rest of the test invalid, consider using `require.NoError` to record the failure and stop the test instead of continuing.

E.g.
```go
require.NoError(mt, err, "MarshalAppend error")
```

That suggestion applies for all "nil error" assertions that use `assert`.

File: mongo/integration/client_side_encryption_prose_test.go [hunk 9]
```diff
@@ -35,6 +36,8 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 	"go.mongodb.org/mongo-driver/mongo/writeconcern"
 	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/mongocrypt"
+	mongocryptOpts "go.mongodb.org/mongo-driver/x/mongo/driver/mongocrypt/options"
```
Comment (matthewdale) [inline mongo/integration/client_side_encryption_prose_test.go:40]: Optional: Go package names are conventionally all lowercase. Using a camel-case package alias could be confusing because it looks like a variable or function.

E.g.
```go
mongocryptopts "go.mongodb.org/mongo-driver/x/mongo/driver/mongocrypt/options"
```

**Review by matthewdale — APPROVED** [review]
A few suggestions, but looks good 👍
